processing/src/main/java/io/druid/query/aggregation/FirstAggregatorFactory.java

```diff
@@ -0,0 +1,338 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.aggregation;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.primitives.Doubles;
+import com.google.common.primitives.Longs;
+import com.metamx.common.IAE;
+import com.metamx.common.StringUtils;
+import io.druid.collections.SerializablePair;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.column.Column;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+public class FirstAggregatorFactory extends AggregatorFactory
+{
+  private static final byte CACHE_TYPE_ID = 0x10;
+  private static final String TYPE_LONG = "long";
+  private static final String TYPE_DOUBLE = "double";
+
+  private final String fieldName;
+  private final String name;
+  private final String value;
+
+  @JsonCreator
+  public FirstAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") final String fieldName,
+      @JsonProperty("value") String value
+  )
+  {
+    Preconditions.checkNotNull(name, "Must have a valid, non-null aggregator name");
+    Preconditions.checkNotNull(fieldName, "Must have a valid, non-null fieldName");
+    Preconditions.checkArgument(
+        value.equals(TYPE_LONG) || value.equals(TYPE_DOUBLE),
+        "Must have a valid, non-null type"
+    );
+
+    this.name = name;
+    this.fieldName = fieldName;
+    this.value = value;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    if (value.equals(TYPE_LONG)) {
+      return new LongFirstAggregator(
+          name, metricFactory.makeLongColumnSelector(fieldName),
+          metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME)
+      );
+    } else if (value.equals(TYPE_DOUBLE)) {
+      return new DoubleFirstAggregator(
+          name, metricFactory.makeFloatColumnSelector(fieldName),
+          metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME)
+      );
+    }
+    throw new IAE("undefined type");
+  }
+
+  @Override
+  public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+  {
+    if (value.equals(TYPE_LONG)) {
+      return new LongFirstBufferAggregator(
+          metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME),
+          metricFactory.makeLongColumnSelector(fieldName)
+      );
+    } else if (value.equals(TYPE_DOUBLE)) {
+      return new DoubleFirstBufferAggregator(
+          metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME),
+          metricFactory.makeFloatColumnSelector(fieldName)
+      );
+    }
+    throw new IAE("undefined type");
+  }
+
+  @Override
+  public Comparator getComparator()
```

Comment:
The comparator from `getComparator` is used for ranking by metric in topNs and groupBys; we probably would rather rank by the actual last value rather than by the timestamp.